sdk/opa.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package sdk
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/open-policy-agent/opa/bundle"
+	initload "github.com/open-policy-agent/opa/internal/runtime/init"
+	"github.com/open-policy-agent/opa/internal/uuid"
+	"github.com/open-policy-agent/opa/loader"
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/plugins/discovery"
+	"github.com/open-policy-agent/opa/plugins/logs"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/server"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+)
+
+// OPA represents an instance of the policy engine.
+type OPA struct {
+	manager     *plugins.Manager
+	configBytes []byte
+	paths       []string
+	filter      loader.Filter
+	bundleMode  bool
+}
+
+// ConfigFile sets the configuration file to use on the OPA instance.
+func ConfigFile(fileName string) func(*OPA) error {
+	return func(opa *OPA) error {
+		bs, err := ioutil.ReadFile(fileName)
+		if err != nil {
+			return err
+		}
+		opa.configBytes = bs
+		return nil
+	}
+}
+
+// Config sets the configuration to use on the OPA instance.
+func Config(config string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.configBytes = []byte(config)
+		return nil
+	}
+}
+
+// Paths sets the paths to search for policy, data files and bundles
+func Paths(paths []string) func(*OPA) error {
+	return func(opa *OPA) error {
+		opa.paths = paths
+		return nil
+	}
+}
+
+// New returns a new OPA object.
+func New(opts ...func(*OPA) error) (*OPA, error) {
+	opa := &OPA{}
+
+	for _, opt := range opts {
+		if err := opt(opa); err != nil {
+			return nil, err
+		}
+	}
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var pluginInitializers []func(*plugins.Manager)
+
+	if opa.paths != nil && len(opa.paths) > 0 {
+		loaded, err := initload.LoadPaths(opa.paths, opa.filter, opa.bundleMode, nil, false)
+		if err != nil {
+			return nil, fmt.Errorf("load error %v", err)
+		}
+		pluginInitializers = append(pluginInitializers, plugins.InitBundles(loaded.Bundles))
+		pluginInitializers = append(pluginInitializers, plugins.InitFiles(loaded.Files))
+	}
+
+	opa.manager, err = plugins.New(opa.configBytes, id, inmem.New(), pluginInitializers...)
```

Comment:
This enforces a hard 1:1 relation between manager and storage. If you want to share the storage across multiple pluginmanagers that should be another option to pass.

Reply:
Fair point! I'll change the `storage` object to be injectable for that purpose.

Reply:
Fixed.